Analysis/DataModel/include/Analysis/JetFinder.h

```diff
@@ -0,0 +1,234 @@
+      jetEtaMax = etaMax - jetR;
+    }
+
+    //selGhosts =fastjet::SelectorRapRange(ghostEtaMin,ghostEtaMax) && fastjet::SelectorPhiRange(phiMin,phiMax);
+    //ghostAreaSpec=fastjet::GhostedAreaSpec(selGhosts,ghostRepeatN,ghostArea,gridScatter,ktScatter,ghostktMean);
+    ghostAreaSpec = fastjet::GhostedAreaSpec(ghostEtaMax, ghostRepeatN, ghostArea, gridScatter, ktScatter, ghostktMean); //the first argument is rapidity not pseudorapidity, to be checked
+    jetDef = fastjet::JetDefinition(algorithm, jetR, recombScheme, strategy);
+    areaDef = fastjet::AreaDefinition(areaType, ghostAreaSpec);
+    selJets = fastjet::SelectorPtRange(jetPtMin, jetPtMax) && fastjet::SelectorEtaRange(jetEtaMin, jetEtaMax) && fastjet::SelectorPhiRange(jetPhiMin, jetPhiMax);
+    jetDefBkg = fastjet::JetDefinition(algorithmBkg, jetBkgR, recombSchemeBkg, strategyBkg);
+    areaDefBkg = fastjet::AreaDefinition(areaTypeBkg, ghostAreaSpec);
+    selRho = fastjet::SelectorRapRange(bkgEtaMin, bkgEtaMax) && fastjet::SelectorPhiRange(bkgPhiMin, bkgPhiMax); //&& !fastjet::SelectorNHardest(2)    //here we have to put rap range, to be checked!
+  }
+
+  /// Sets the background subtraction estimater pointer
+  void setBkgE()
+  {
+    if (bkgSubMode == BkgSubMode::rhoAreaSub || bkgSubMode == BkgSubMode::constSub) {
+      bkgE = decltype(bkgE)(new fastjet::JetMedianBackgroundEstimator(selRho, jetDefBkg, areaDefBkg));
+    } else {
+      if (bkgSubMode != BkgSubMode::none)
+        LOGF(ERROR, "requested subtraction mode not implemented!");
+    }
+  }
+
+  /// Sets the background subtraction pointer
+  void setSub()
+  {
+    //if rho < 1e-6 it is set to 1e-6 in AliPhysics
+    if (bkgSubMode == BkgSubMode::rhoAreaSub) {
+      sub = decltype(sub){new fastjet::Subtractor{bkgE.get()}};
+    } else if (bkgSubMode == BkgSubMode::constSub) { //event or jetwise
+      constituentSub = decltype(constituentSub){new fastjet::contrib::ConstituentSubtractor{bkgE.get()}};
+      constituentSub->set_distance_type(fastjet::contrib::ConstituentSubtractor::deltaR);
+      constituentSub->set_max_distance(constSubRMax);
+      constituentSub->set_alpha(constSubAlpha);
+      constituentSub->set_ghost_area(ghostArea);
+      constituentSub->set_max_eta(bkgEtaMax);
+      constituentSub->set_background_estimator(bkgE.get()); //what about rho_m
+    } else {
+      if (bkgSubMode != BkgSubMode::none)
+        LOGF(ERROR, "requested subtraction mode not implemented!");
+    }
+  }
+
+  /// Performs jet finding
+  /// \note the input particle and jet lists are passed by reference
+  /// \param inputParticles vector of input particles/tracks
+  /// \param jets veector of jets to be filled
+  /// \return ClusterSequenceArea object needed to access constituents
+  fastjet::ClusterSequenceArea findJets(std::vector<fastjet::PseudoJet>& inputParticles, std::vector<fastjet::PseudoJet>& jets) //ideally find a way of passing the cluster sequence as a reeference
+  {
+    setParams();
```

Comment:
What's the reason to do this for every call of the actual jet finding? Isn't it sufficient to have the settings initialized once?

Reply:
the reason for the param initialisation is that thee user may wish to change certain variables after initialising an instance of the class (for example etaMax for run5 studies). If the params are not initialised again then the corresponding selectors and other variables like the jet radius which depend on such parameters would not be updated. I have added eta and phi options directly in the constructor but this holds for example for any of the other variables going into the selectors. That was the logic but perhaps there is a more elegant way to achieve this

Reply:
I can see that you would want to change the parameters after initialising the class, but not between running the jet finding. In case you want jet finding with different settings in the same event, wouldn't it be more logical to setup two instances of the jet finding class? Maybe I am missing something here?

Reply:
well there can be cases of jet correlations with a trigger object where the phi might want to be changed each time in accordance with the recoil window from the trigger. This can of course also be done after the jet finding too but can also be done via the selector options.